src/components/panel/panel.js

```diff
@@ -1636,6 +1685,91 @@ MdPanelRef.prototype._animateClose = function() {
   });
 };
 
+/**
+ * Registers a listener with the panel. The callback should return a promise,
+ * which will allow the action to continue when it gets resolved, or will
+ * prevent an action if it is rejected.
+ * @param {string} type Type of listener.
+ * @param {!angular.$q.Promise<!any>} callback Callback to be registered.
+ * @returns {!MdPanelRef}
+ */
+MdPanelRef.prototype.registerListener = function(type, callback) {
```

Comment:
Perhaps this ties in with a more generic use for the service introduced with #9514? I went with this approach mainly because of the spec by @ErinCoughlan in #9557. 

As an aside, I was initially considering something similar to the browser API where you could do `event.preventDefault` to stop the panel from closing, but using promises should be more flexible since it allows you to determine the close condition asynchronously.